x-pack/plugin/analytics/src/main/java/org/elasticsearch/xpack/analytics/mapper/HistogramFieldMapper.java

```diff
@@ -491,76 +514,85 @@ protected SyntheticSourceMode syntheticSourceMode() {
+    }
 
-            @Override
-            public DocValuesLoader docValuesLoader(LeafReader leafReader, int[] docIdsInLeaf) throws IOException {
-                BinaryDocValues docValues = leafReader.getBinaryDocValues(fieldType().name());
-                if (docValues == null) {
-                    // No values in this leaf
-                    binaryValue = null;
-                    return null;
-                }
-                return docId -> {
-                    if (docValues.advanceExact(docId)) {
-                        binaryValue = docValues.binaryValue();
-                        return true;
-                    }
-                    binaryValue = null;
-                    return false;
-                };
-            }
+    private class HistogramSyntheticFieldLoader implements CompositeSyntheticFieldLoader.SyntheticFieldLoaderLayer {
+        private final InternalHistogramValue value = new InternalHistogramValue();
+        private BytesRef binaryValue;
 
-            @Override
-            public boolean hasValue() {
-                return binaryValue != null;
-            }
+        @Override
+        public Stream<Map.Entry<String, SourceLoader.SyntheticFieldLoader.StoredFieldLoader>> storedFieldLoaders() {
+            return Stream.of();
+        }
 
-            @Override
-            public void write(XContentBuilder b) throws IOException {
-                if (binaryValue == null) {
-                    return;
+        @Override
+        public SourceLoader.SyntheticFieldLoader.DocValuesLoader docValuesLoader(LeafReader leafReader, int[] docIdsInLeaf)
+            throws IOException {
+            BinaryDocValues docValues = leafReader.getBinaryDocValues(fieldType().name());
+            if (docValues == null) {
+                // No values in this leaf
+                binaryValue = null;
+                return null;
+            }
+            return docId -> {
+                if (docValues.advanceExact(docId)) {
+                    binaryValue = docValues.binaryValue();
+                    return true;
                 }
-                b.startObject(simpleName());
+                binaryValue = null;
+                return false;
+            };
+        }
 
-                value.reset(binaryValue);
-                b.startArray("values");
-                while (value.next()) {
-                    b.value(value.value());
-                }
-                b.endArray();
+        @Override
+        public boolean hasValue() {
+            return binaryValue != null;
+        }
 
-                value.reset(binaryValue);
-                b.startArray("counts");
-                while (value.next()) {
-                    b.value(value.count());
-                }
-                b.endArray();
+        @Override
+        public void write(XContentBuilder b) throws IOException {
+            if (binaryValue == null) {
+                return;
+            }
+            b.startObject();
```

Comment:
Why was this changed from `b.startObject(simpleName());` ?

Reply:
This is because composite loader writes that now. It is possible that there are malformed values so this is now not an object but an array.